### ISSUE TITLE
support and eventually upgrade to an alternate jobserver protocol standard that does not leak threads, in collaboration with zig

### ISSUE THREAD

**Issue by andrewrk** [rust-lang/cargo#14102]
### Problem

The GNU make jobserver protocol leaks threads if a process dies while holding a token. For example:

1. Parent process creates the pipe and puts 2 tokens into the pipe, then spawns children A and B.
2. Child Process A reads a token and starts doing work, but then crashes before putting the token back.
3. Child Process B has 100 tasks to do, but only proceeds to do them using 1 logical core, leaving an entire core idle because it has been "leaked" by Child Process A.

It could get even worse - if another leak occurred in this scenario, the entire process tree would deadlock.

The fundamental problem is that the protocol relies on a process manually releasing resources before it dies, rather than relying on kernel-managed resources that are automatically released when the process dies.

### Proposed Solution

Let's work together on a standard protocol that does not have this flaw, and then pressure GNU make, Ninja, and other third party projects to support it in future versions.

Here is an initial proposal (copied from https://github.com/ziglang/zig/issues/20274). This is something I thought of today and have not evaluated yet. I'm happy to workshop it with the Cargo team.

Root process opens a temporary file (perhaps with with `shm_open`) and writes N bytes to the file, where N is thread pool size. This file descriptor is passed down the process tree. Each byte in the file represents an available logical core. Each thread in each process's thread pool holds an advisory record lock (`fnctl(F_SETLKW)`) on the corresponding byte while working. Advisory record locks are automatically released by the kernel when a process dies.

In theory this is exactly the same overhead as reading and writing byte-sized tokens to a pipe (1 syscall each for lock and unlock), the difference being the resource is automatically released by the kernel on process death.

The idea would be to deprecate the legacy GNU make jobserver protocol and try to standardize various open source projects on the better protocol. In the beginning, it could be merely *supported* when the parent process implements this new standard, and have a flag for choosing it when Cargo is the root process in the tree. Later on, if the new standard achieves widespread adoption, then the default could change.

### Notes

If there is interest, I will follow up here with performance data points and other observations as I evaluate this strategy in the Zig project.

Requested label from triagers: `A-jobserver`

Happy hacking!

**Comment by weihanglo** [comment]
cc @NobodyXu you might be interested.

**Comment by NobodyXu** [comment]
Interesting proposal but I have a few questions:
 - how would you find a byte to lock?
 - how would you block a process until a token is available?
 - how to implement this on windows?

Sorry if I sound discouraging, but I am just reading your proposals and is thinking about how to implement it.

**Comment by andrewrk** [comment]
> * how would you find a byte to lock?

Each process that wants to participate spawns a thread pool of size N where N is the byte size of the file. The threads are numbered sequentially from 0..N, each corresponding to a byte in the file.

> * how would you block a process until a token is available?

Lock:

```c
struct flock options = {
    .l_type = F_WRLCK,
    .l_whence = SEEK_SET,
    .l_start = thread_index,
};
fcntl(fd, F_SETLKW, &options);
```

Unlock:

```c
struct flock options = {
    .l_type = F_UNLCK,
    .l_whence = SEEK_SET,
    .l_start = thread_index,
};
fcntl(fd, F_SETLKW, &options);
```

> * how to implement this on windows?

Out of scope. A completely different strategy probably.

**Comment by Hello71** [comment]
This scheme seems suspiciously quadratic in memory usage, with a high constant factor. From https://unix.stackexchange.com/questions/708104/what-is-the-minimum-amount-of-memory-required-for-starting-a-process-on-a-linux, each thread costs at least one 10 KB `task_struct`, plus at least 4 KiB stack, so 128 jobs * 128 threads/job * ~16 KiB/thread = ~262 MiB. Even if memory is cheap, that's way too much to spend on jobserver synchronization.

Also, the kernel is probably not optimized for very large numbers of threads waiting to lock different bytes in the same file. I wouldn't be surprised if there's a linked list of all waiters on each file, because it would use less memory than a tree in normal usage.

**Comment by andrewrk** [comment]
Hmm, I'm not following your math. Is this example a computer that has 128 logical cores? In this case it would be 128 threads per process. So, 16 KiB/thread * 128 threads = 2 MiB per process. In a more typical case of 32 logical cores it would be 16 KiB/thread * 32 threads = 512 KiB per process.

**Comment by andrewrk** [comment]
> Also, the kernel is probably not optimized for very large numbers of threads waiting to lock different bytes in the same file. I wouldn't be surprised if there's a linked list of all waiters on each file, because it would use less memory than a tree in normal usage.

It appears to me this hunch is correct. In the Linux source code, for example, in fs/locks.c, `posix_lock_inode`, I can see that it uses `list_for_each_entry` to iterate over locks, and based on a cursory glance it seems like that list is per-file.

It's disappointing. I haven't found any other primitives that might satisfy this use case. 24 million lines of C code in this source tree, 1.3 million commits, 23 years of heavy development, and yet we as users can't even batch work across a process tree correctly.

**Comment by andrewrk** [comment]
Alright, never mind, there was a solution right in front of my face. Here's an alternate scheme:

Root process listens on a unix domain socket. That fd is passed down the process tree. To get a thread token a child process connects to the socket. To return the token, disconnect. The root process only accepts N concurrent connections and stops accepting when it's maxed out.

When a child process dies, the OS causes the connection to end, allowing the root process to accept a new connection.

**Comment by NobodyXu** [comment]
Yeah the unix socket sounds plausible, it's also cross platform and can be doable on windows and other OSes.

It adds a bit more work to the "root process", where as the make jobserver doesn't need a server, but I think this is a reasonable design and in practice you always need a "root process" to keep it alive anyway.

The socket connection can be either blocking or non-blocking, so usable in sync or async function.

Compared to [jobserver](https://docs.rs/jobserver/latest/jobserver/struct.Client.html), it can't get the available tokens, though I guess that can be implemented via an extension by sending message to the server.

Also, I think the address of the unix socket would need to be passed as environment variables?
E.g. `JOBSERVER=unix-socket:...`

**Comment by mlugg** [comment]
> Compared to [jobserver](https://docs.rs/jobserver/latest/jobserver/struct.Client.html), it can't get the available tokens, though I guess that can be implemented via an extension by sending message to the server.

What's the actual use case for such a feature? I'm not claiming there isn't one, I just don't immediately see it. AFAICT, any process spawning jobs can just use a thread pool capped at the number of logical cores; perhaps there will be extra threads blocked at all times, but that shouldn't be a problem.

> Also, I think the address of the unix socket would need to be passed as environment variables? E.g. `JOBSERVER=unix-socket:...`

Yep, agreed.

**Comment by NobodyXu** [comment]
> What's the actual use case for such a feature?

Hmmm I can't think of one, so yeah probably isn't that important.

**Comment by mlugg** [comment]
> Suppose that you spawn a new task after getting a token from jobserver protocol, that task has an implicit token. for its main thread.

When the parent spawns its child, could the child directly inherit the FD for the connection acquired by the parent? Then, the parent could close their FD after the fork, and the child now "owns" that socket connection, and it will be automatically released when the child exits. We still need logic for this (to ensure the child inherits the FD and the parent immediately closes it), but I think having it in child spawning is mildly better than requiring every child to handle this.

**Comment by NobodyXu** [comment]
Yeah I think that could work.

Having the fd in child would also have another benefit:

The job token will be released whenever the child exits, instead of whenever the one that acquires the job token has reaped the child or itself exits.

**Comment by bjorn3** [comment]
> When the parent spawns its child, could the child directly inherit the FD for the connection acquired by the parent?

That re-introduces the problem that the new named pipe jobserver protocol of make solves which the old anonymous pipe protocol suffered from: You are forced to initialize the jobserver client before any code runs that could open file descriptors. Otherwise you might misinterpret a locally opened file descriptor as an inherited one originating from the jobserver, which violates IO safety. This also necessitates `jobserver::Client::from_env` being unsafe.

**Comment by NobodyXu** [comment]
Hmmm yeah that's true

**Comment by andrewrk** [comment]
I haven't been able to actually get this scheme to work in practice, on Linux: https://gist.github.com/andrewrk/d8014959638ff67ff87a183b639304e2

connect() on an un-accepted connection does not block (?)
poll(POLLOUT) on a connected, un-accepted connection does not block (??)

**Comment by joshtriplett** [comment]
@andrewrk wrote:
> Alright, never mind, there was a solution right in front of my face. Here's an alternate scheme:
> 
> Root process listens on a unix domain socket. The address is passed down the process tree via environment variable. To get a thread token a child process connects to the socket. To return the token, disconnect. The root process only accepts N concurrent connections and stops accepting when it's maxed out.
> 
> When a child process dies, the OS causes the connection to end, allowing the root process to accept a new connection.

This seems like a great solution that's reasonably portable; it'd work on any UNIX, and it'd work on modern Windows.

@andrewrk wrote:
> I haven't been able to actually get this scheme to work in practice, on Linux: https://gist.github.com/andrewrk/d8014959638ff67ff87a183b639304e2
> 
> connect() on an un-accepted connection does not block (?) poll(POLLOUT) on a connected, un-accepted connection does not block (??)

It'd be an extra syscall, but one thing that *should* work portably on any UNIX system would be to connect and then read one byte (whether blocking or non-blocking), and have the jobserver always write a single byte to each socket after accepting it.

If that turns out to be the only way to get this to work, then another possible protocol, which might be faster for multi-job processes but marginally slower for single-job processes, would be for clients of this protocol to first *write* a byte to the connected socket for each job it *wants* to start, and *then* read a byte per job. This would allow connecting a socket *once* and then requesting multiple jobs from it, rather than connecting a socket per job. For a multi-job process, that reduces the overhead to a `write` and `read` per job (both of which may be possible to batch), and it doesn't require the overhead of updating the fd table repeatedly. But for a single-job process, this would mean one *more* syscall (a `write`) on top of what's now already a `socket`/`connect`/`read`.

It's unfortunate that this protocol already requires 3 syscalls per job. This isn't *substantially* worse than the current jobserver protocol (which requires `open` and `read` and a subsequent `write`) or the old one (inherit fd, do a `read` and subsequent `write`). And I don't see any obvious way to do *substantially* better, even without taking portability into account, while preserving the property of automatically releasing on close. (I can see ways to do slightly better but they'd add substantial complexity; for instance, sendmsg of a pipe.)

**Comment by andrewrk** [comment]
I am observing one thread blocking on connect(), and then another thread calls shutdown() on the same sockfd, the first thread continues to block on the connect syscall: https://gist.github.com/andrewrk/bf46cff35d34692558563f75bc8d0e62

Any ideas on how to solve this? The problem is on deinitialization; the main thread needs a way to wake up the worker threads that are blocking on connect().

**Comment by joshtriplett** [comment]
@andrewrk Based on the earlier experiment, I had thought `connect` never blocked in the first place, which is why we need the `read`.

Investigating further, it looks like eventually `connect` *will* block. depending on the backlog size specified in `listen`. (Looking at the zig PR thread, it looks like you found something similar.)

For instance, if you have a backlog of `2`, and your listener accepts only one connection and loops forever (never accepting other connections), the first connect will succeed (being accepted), the next *3* (?!) will successfully `connect` and block in `read`, and after that, additional `connect` attempts will block.

Likewise, with a backlog of `1`, the first connection succeeds (being accepted), the next *2* (?!) will successfully `connect` and block in `read`, and after that, additional `connect` attempts will block.

This behavior seems weird enough that it probably can't be counted on, and I'd be *shocked* if it doesn't vary by platform, so I'm not sure if counting on it is a good idea. I'm not sure if it's viable to count on setting the backlog to a large number (in any case you shouldn't set it to more than SOMAXCONN), nor should you count on any particular backlog behavior from the jobserver (since many implementations are unlikely to change the default). I think you're going to have to allow for *either* `connect` *or* `read` blocking.

From what I can tell, the behavior of non-blocking `connect` is *also* both finicky and non-standard here; I don't think you're likely to be able to use `poll` to wait for connect to succeed. On Linux at least, non-blocking UNIX sockets just fail `connect` with `EAGAIN` rather than returning `EINPROGRESS`, so you I don't think you *can* `poll` for completion of `connect`. This seems like a substantial deficiency. I haven't found *any* reasonable way to do a non-blocking `connect` to a UNIX socket on Linux; there are various unreasonable ways that require jumping through substantial hoops (e.g. *requiring* `io_uring`, or spawning a thread/process).

You could use the sockopt `SO_SNDTIMEO`, which is standardized by POSIX, but I'm not sure all platforms apply that timeout to `connect` and not just `send`/`write`. In any case, this would cause you to have to repeatedly wake up when the timeout expires and re-invoke `connect`, which seems wasteful, and it'd incur a delay before you can kill worker threads. That *might* be acceptable, though, since you expect most blocks to occur on `read` (for which `poll` works) rather than `connect`. But in any case, you'd have to verify that it applies to `connect` on all targets.

I don't think you can portably send a signal to a specific thread (rather than a process), leaving aside all the other reasons not to want to use signals. `pthread_cancel` similarly seems like a bad idea.

I expect the behavior of closing the fd (or race-free equivalent like `dup2`ing `/dev/null` over the fd) would be finicky and non-standard. From reading the Linux code, I *think* that if `connect` has reached the point of waiting to connect, it's just going to keep waiting and not care about its socket having been closed.

You could move the blocking from the worker threads themselves to a separate thread that handles making the connections, which would allow you to terminate the worker threads (since they could wait on something like a channel), but that just moves the problem to the connection thread.

If you want something to be able to block in either `connect` or `read`, but you *also* want to be able to bail out and end the worker pool by some means other than your process ending, I don't have anything else obvious to suggest; non-blocking `connect` doesn't work on UNIX sockets, and blocking connect can't easily be terminated.

**Comment by joshtriplett** [comment]
@andrewrk Noting another possible protocol, which would be conceptually simple but substantially more complex to implement and deal with: a client opens a named pipe and reads a byte, and expects the byte to come with a file descriptor via `SCM_RIGHTS`; rather than writing a byte back to the pipe to indicate you're done, close the file descriptor. Upsides: all of that should work fine in a non-blocking manner, `poll` will Just Work, and it's just two syscalls per job (`recvmsg`, `close`). Downside: UNIX socket ancillary data is *painful* to work with in many languages (e.g. Rust has no native support in std yet, you'd have to use something like `rustix::net::recvmsg`), whereas "connect to UNIX socket and read a byte" is trivial from any language...as long as you don't need it to be non-blocking.

**Comment by joshtriplett** [comment]
A correction: it looks like signals might be viable after all. https://www.man7.org/linux/man-pages/man3/pthread_kill.3.html exists. You could install a signal handler, invoke `pthread_kill`, the handler will run on that thread, the handler can set a flag, and after the handler returns, `connect` should return with `EINTR` and the code that invoked `connect` can check the flag.

**Comment by andrewrk** [comment]
Update: I'm still interested in this but have put the project on ice for the time being. If y'all end up looking into it and figuring out something that works well, Zig will follow Cargo's lead. For now though, I was unable to come up with something satisfactory with the available primitives.

**Comment by mlugg** [comment]
I spent some time looking at this yesterday, and I came up with a few schemes which seem hopeful.

## System V semaphores

What we're actually trying to implement here is just a counting semaphore with thread termination safety. It turns out that an old System V semaphore primitive (not to be confused with POSIX semaphores!) solves this *perfectly*. When you increment or decrement the value of one of these semaphores (with `semop`), you can specify the flag `SEM_UNDO` to signal to the kernel that this modification should be reverted when the process exits.

So in our case, we would specify `SEM_UNDO` on both the decrement (acquiring a token) and the increment (releasing a token), and they neatly cancel out so that everything Just Works---when this process exits it's as if it never modified the semaphore. Note that this isn't doing something inefficient in the kernel like storing a list of all these modifications; it's just incrementing/decrementing a per-process "adjustment" value which will get added into the semaphore upon process termination.

The calls in the server would look like this:
```c
/* Despite the name `IPC_PRIVATE`, the return value here is a System V IPC handle
 * (not an FD!) which can be safely passed straight to child processes. */
const int sem_id = semget(IPC_PRIVATE, 1, 0777);
semctl(sem_id, 0, SETVAL, num_tokens);
/* Then pass `sem_id` to children somehow (probably via an env var). */
```
And in a worker:
```c
/* Somehow acquire `sem_id` as exposed by the server (e.g. read an env var) */
const int sem_id = ...;

/* To acquire a token: */
{
    struct sembuf op_buf = { .sem_num = 0, .sem_op = -1, .sem_flg = SEM_UNDO };
    semop(sem_id, &op_buf, 1);
}

do_some_work(); /* we have a token, use it */

/* To release a token: */
{
    struct sembuf op_buf = { .sem_num = 0, .sem_op = 1, .sem_flg = SEM_UNDO };
    semop(sem_id, &op_buf, 1);
}
```

I've tested this concept with a simple implementation in Zig, and it does seem to work nicely. At least on Linux, each `semop` is a single syscall, so acquiring and releasing a token are one syscall each. The kernel implementation ([here](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/ipc/sem.c)) looks totally reasonable, so I have no performance concerns.

It seems to me that where available, this is by *far* the best option we've seen yet. The only slight concern I have is support for this primitive on different platforms; POSIX semaphores (which do *not* have the process termination safety properties we need) have largely replaced System V semaphores in common usage, so it's possible some implementations don't support the latter properly or have buggy implementations. But perhaps I'm being too pessimistic, and they still work fine on many OSes; this needs investigating.

## `AF_UNIX` + `SOCK_DGRAM` + `SCM_RIGHTS` + `pipe2`

I came up with this idea first before realising that System V semaphores were viable. The scheme I'm about to describe seems to be pretty clearly *strictly worse* than the above one, so is probably only worth thinking about if System V semaphores turn out to be problematic.

The idea is to model each token as a pipe, sending the write-end to a client through a Unix domain socket, and using the fact that the server can detect when the pipe is broken (e.g. as a `POLLERR` from a `poll` call).

Server logic:
* Open an `AF_UNIX` `SOCK_DGRAM` socket somewhere (presumably we pass the path down to children via an env var).
  * The reason we're using a datagram socket is purely to optimize syscalls: this will allow clients to receive data without needing to `connect`.
* For `0..num_tokens`, call `pipe2` to get pairs of FDs.
* Using `sendmsg` with `SCM_RIGHTS`, send the write-end of each pipe to the socket
* Close the write-ends of all pipes.
* `poll` on the read-ends of all pipes.
* Clients release tokens by closing the write-end of a pipe. This causes the server to see `POLLERR` from the corresponding read-end, because the write-end of the pipe is no longer open in any process.
  * When the server sees this, it must create a *new* pipe for the token with `pipe2` and send that pipe's write end back over the socket to let another client work.

Corresponding client logic:
* Determine the path to the server's socket (probably from an env var).
* To acquire a token, call `recvmsg` on the socket. You will get a datagram with an attached FD; this represents your token.
* To release a token, close the attached FD. (Of course, if the process terminates unexpectedly, the FD will automatically close.)

On the client end, this is prety simple: 1 syscall to acquire a token, and 1 syscall to release it. The server isn't looking so great though; per acquire-release pair, the server does `pipe2`, `sendmsg` (transferring the write end of the pipe), `close` (on the write end of the pipe), `poll` (to wait for release), `close` (on the read end of the broken pipe). That's pretty bad. I suppose it would at least be fairly easy to use batching mechanisms like io_uring on the server here. Like I said, this seems pretty terrible compared to the System V semaphore scheme.

**Comment by mlugg** [comment]
Based on the System V semaphore idea above, and some collaboration with other Zig team members on a design for Windows, I've put together a concrete proposed protocol which I've called the "Robust Jobserver". You can find it here: https://codeberg.org/mlugg/robust-jobserver/src/branch/main/spec.md

I'd be interested to hear if anyone has thoughts or concerns regarding the specification as written. If not, I intend to work on supporting it in the Zig compiler and build system at some point soon.

**Comment by NobodyXu** [comment]
IMHO the windows implementation using named pipe, how do we know the exact name for the jobserver?

Do we have to guess?

**Comment by mlugg** [comment]
I don't completely understand the question (sounds like something may have been lost in translation?). To be clear: the server picks any pipe name it likes, and communicates it to clients through the `ROBUST_JOBSERVER` environment variable. If that variable's value is `win32pipe:foo`, then the name of the pipe is `\\.\foo`.

**Comment by NobodyXu** [comment]
But there could be multiple client, and for windows you'd have one named pipe per token, so for each client, what is the name of the named pipe?

It can't be a fixed name right?

**Comment by mlugg** [comment]
Windows has a concept called named pipe "instances", which is basically a bunch of pipes with the same name. What Windows calls a named pipe is actually much closer to what us POSIX-minded folk would consider a local socket: it's (potentially) bidirectional, and supports arbitrarily many connections (through this "instances" concept), with a "server" accepting connections from "clients". So, yes, all clients connect to the same pipe name. For more information you can read Microsoft's documentation here: https://learn.microsoft.com/en-us/windows/win32/ipc/named-pipes

For the record, we already (before I wrote up the spec) successfully wrote and tested PoC implementations for both of the listed communication methods. The `win32pipe` method has been tested on Windows, and the `sysvsem` method has been tested on Linux and macOS.

**Comment by NobodyXu** [comment]
Thanks for the explanation!

I'm not really familiar with Windows so I was confused by this, it'd be great if the doc explained that to people not familiar with Windows.

Just curious, is it possible to avoid having one thread for each token on windows? Is it possible to have something "async"?

**Comment by mlugg** [comment]
Yes; one thread per token is easiest to explain, hence why the spec describes it, but our PoC implementation used just one server thread with asynchronous pipe operations and [Asynchronous Procedure Calls](https://learn.microsoft.com/en-us/windows/win32/sync/asynchronous-procedure-calls).

**Comment by NobodyXu** [comment]
That looks pretty nice, looking forward to a rust implementation in rust-lang/jobserver-rs

**Comment by epage** [comment]
Unsure how relevant this is but sounds like Gentoo folks are looking at a system-level jobserver: https://blogs.gentoo.org/mgorny/2025/11/30/one-jobserver-to-rule-them-all/

**Comment by mlugg** [comment]
@epage regarding system-wide usage, note that my protocol is specifically designed to allow a job server to advertise itself from arbitrarily far up the process tree---take a look at the last section: https://codeberg.org/mlugg/robust-jobserver/src/branch/main/spec.md#multiple-servers. The use case I had in mind there was correctly managing jobs across a whole bunch of different compilations on a CI machine, which it sounds like is along the same lines as the Gentoo folks were thinking.

**Comment by andrewrk** [comment]
@mlugg maybe it would help to link to the PoC code?

**Comment by mlugg** [comment]
@andrewrk indeed, I was kind of procrastinating on getting it to a state where I'm happy to show it---but I'm there now!

The Zig implementation I've got can be found on the `jobserver` branch of the Zig repo: https://codeberg.org/ziglang/zig/commits/branch/jobserver. It's much more polished than my original PoC---error handling should all be correct, and it implements both the POSIX-ish communication method (`sysvsem`) and the Windows-specific one (`win32pipe`; this also works correctly under Wine). The penultimate commit on that branch is the one with the actual implementation; it's pretty straightforward (even though the Windows implementation is an efficient one based on APCs, rather than the naive threaded approach).

Based on the final commit of the branch (just a simple usage test) it seems to work as expected: the output shows (on my system with nproc=16) that exactly 16 workers are running at all times (when one finishes, another starts). I've also tried extending that test to have random threads occasionally crash, and can confirm that all tokens owned by that process are released and another process can immediately start running (like we want!). Put simply, I've not observed any problems so far!

I expect we'll move towards including this in the Zig standard library at some point soon-ish, so it'd be great for y'all to take a look at this so that any concerns or issues can be raised *before* we start trying to actually use it :^)

**Comment by Ext3h** [comment]
As far as only targeting Linux goes, https://wiki.gentoo.org/wiki/Steve was already as stable as it gets while sticking purely to the GNU Make jobserver protocol.

The problem that it didn't solve: It still only works within a single file system, which is more limiting than anticipated.

@mlugg While it's nice to see that the jobserver is now robust enough to survive thread termination, it has still inherited the flaw of the original GNU Make jobserver that it's fundamentally bound to a single "system". The reality out there is that with modern multi-platform development, you rarely have just one OS side namespace, or even OS running on the same hardware platform. And with RAM becoming premium, exactly tailoring your loads even in a heavily virtualized and containerized environment has become a common requirement.

Comparing `steve` to your solution, you are running into quite the similar issue - you went for two rather exotic and fundamentally incompatible mechanisms to support different platforms. Both of which can't pass OS boundaries. When the real need would be to finally have a common jobserver protocol that works flawlessly over OS boundaries, strictly using only technologies that are available cross-platform and have common, ready made abstractions in every relevant language out there.

The only realistically feasible approach I can see here is actually going for a real server-client infrastructure. GNU Make jobserver was ***close*** to getting it right, in the sense that you could already try to run it over domain sockets instead of pipes. Didn't quite make the turn to allow TCP (over IP) sockets though, and didn't realize the need for REQUESTS before leasing tokens when using non-shared connections to avoid having to prematurely lease tokens just to fill up a pipe/queue/semaphore.

The additional overhead for establishing a TCP socket would still be negligible, especially considering that ***usually*** only the job dispatcher ends up communicating with the server, and rarely the downstream processes. The syscall overhead for actually sending single-byte messages over TCP locally is still negligible, and is hitting one of the best optimized paths in any OS. While there is a small round-trip delay involved when you have to follow a request->lease  round trip rather than a single syscall to acquire a native semaphore, this is probably a cost that should be simply accepted, as it can usually be successfully masked with a very slight over subscription factor.

Please don't try to stick too closely to the performance constraints the original GNU Make jobserver protocol was engineered around. Having multiple logical processors was still considered the exception rather than the normal case, OS side scheduling wasn't exactly mature yet and spending just a few kB of extra memory was still considered wasteful. Things have changed, fundamentally.
